sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Processor/EventProcessorClientLiveTests.cs

```diff
@@ -31,5 +36,831 @@ namespace Azure.Messaging.EventHubs.Tests
     [Category(TestCategory.DisallowVisualStudioLiveUnitTesting)]
     public class EventProcessorClientLiveTests
     {
+        /// <summary>The name of the custom event property which holds a test-specific artificial sequence number.</summary>
+        private static readonly string CustomIdProperty = $"{ nameof(EventProcessorClientLiveTests) }::Identifier";
+
+        /// <summary>The name of the custom event property which holds a test-specific artificial sequence number.</summary>
+        private static readonly string CustomSequenceProperty = $"{ nameof(EventProcessorClientLiveTests) }::Sequence";
+
+        /// <summary>A generator for random values used within the tests.</summary>
+        private readonly Random RandomGenerator = new Random();
+
+        /// <summary>
+        ///   Verifies that the <see cref="EventProcessorClient" /> can read a set of published events.
+        /// </summary>
+        ///
+        [Test]
+        public async Task EventsCanBeReadByOneProcessorClient()
+        {
+            // Setup the environment.
+
+            await using EventHubScope scope = await EventHubScope.CreateAsync(2);
+            var connectionString = TestEnvironment.BuildConnectionStringForEventHub(scope.EventHubName);
+
+            // Test the scenario.
+
+            using var cancellationSource = new CancellationTokenSource();
+            cancellationSource.CancelAfter(TimeSpan.FromMinutes(2));
+
+            var sentCount = 0;
+            var processCount = 0L;
+            var sourceEvents = CreateEvents(50).ToList();
+            var processedEvents = new Dictionary<string, EventData>();
+
+            // Send a set of events.
+
+            await using (var producer = new EventHubProducerClient(connectionString))
+            {
+                foreach (var batch in (await BuildBatchesAsync(sourceEvents, producer, cancellationSource.Token)))
+                {
+                    await producer.SendAsync(batch, cancellationSource.Token);
+
+                    sentCount += batch.Count;
+                    batch.Dispose();
+                }
+            }
+
+            Assert.That(sentCount, Is.EqualTo(sourceEvents.Count), "Not all of the source events were sent.");
+
+            // Attempt to read back the events.
+
+            var completionSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+            var options = new EventProcessorOptions { LoadBalancingUpdateInterval = TimeSpan.FromMilliseconds(250) };
+            var processor = CreateProcessor(scope.ConsumerGroups.First(), connectionString, options: options);
+
+            processor.ProcessErrorAsync += args =>
+            {
+                Assert.Fail($"Processor Error Surfaced: ({ args.Exception.GetType().Name })[{ args.Exception.Message }]");
```

Comment:
Just double checking, but in case the error handler is triggered, will the test actually fail? The handler is triggered with a fire-and-forget call, so if `Assert.Fail` throws, the exception will be lost. Does the test framework keep track of the `Assert.Fail` call?

Reply:
Because we're using an anonymous delegate, it does.   A closure scope is created that captures the reference to `Assert` that exists in the context of the test body when the handler is created/assigned.    When the handler signals `Assert.Fail` the exception context is captured and bubbled.    I can confirm that I had several rounds of failures when writing these attributed to the handler assert.